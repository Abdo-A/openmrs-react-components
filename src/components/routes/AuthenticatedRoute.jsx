import React from 'react';
import { connect } from 'react-redux';
import { Redirect, Route } from 'react-router-dom';

// adapted from: https://reacttraining.com/react-router/web/example/auth-workflow
// will redirect to a Login Page if user is not authenicated
// see sample/App.js for example of usage

const AuthenicatedRoute = props => {

  if (props.session.authenticated === true) {
    return (
      <Route
        component={props.component}
        path={props.path}
      />
    );
  }
  else {
    return (
      <Redirect
        to={{
          pathname: "/login",
          state: { from: props.location }
        }}
      />
    );
  }

};

const mapStateToProps = (state) => {
  return { session: state.openmrs.session };
};

export default connect(mapStateToProps)(AuthenicatedRoute);
