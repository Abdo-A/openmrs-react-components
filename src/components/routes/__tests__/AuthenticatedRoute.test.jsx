import React from 'react';
import MockRouter from 'react-mock-router';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Redirect, Route } from 'react-router-dom';
import AuthenticatedRoute from '../AuthenticatedRoute';

let props, store;
let mountedComponent;

const mockStore = configureMockStore();

const DummyComponent = props => (
  <div>
    <p>
      Sample
    </p>
  </div>
);

const authenicatedRoute = () => {
  if (!mountedComponent) {
    mountedComponent = mount(
      <Provider store={store}>
        <MockRouter>
          <AuthenticatedRoute {...props} />
        </MockRouter>
      </Provider>);
  }
  return mountedComponent;
};

describe('Component: Authenticated Route', () => {
  beforeEach(() => {

    props = {
      component: DummyComponent,
      path: "some_path",
      redirectOnLogin: "/"
    };

    mountedComponent = undefined;
  });

  it('renders properly when session authenicated', () => {

    store = mockStore(
      {
        openmrs: {
          session: {
            authenticated: true
          }
        }
      }
    );

    expect(toJson(authenicatedRoute())).toMatchSnapshot();
    expect(authenicatedRoute().find(Route).length).toBe(1);
    expect(authenicatedRoute().find(Route).props().component).toBe(DummyComponent);
    expect(authenicatedRoute().find(Route).props().path).toBe("some_path");
  });

  it('renders properly when session not authenicated', () => {

    const expectedTo = {
      pathname: "/login",
      state: {
        from: "/"
      }
    };

    store = mockStore(
      {
        openmrs: {
          session: {
            authenticated: false
          }
        }
      }
    );

    expect(toJson(authenicatedRoute())).toMatchSnapshot();
    expect(authenicatedRoute().find(Redirect).length).toBe(1);
    expect(authenicatedRoute().find(Redirect).props().to).toEqual(expectedTo);
  });


});
