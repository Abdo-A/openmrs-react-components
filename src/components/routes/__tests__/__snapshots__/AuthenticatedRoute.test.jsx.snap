// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Authenticated Route renders properly when session authenicated 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MockRouter
    createHref={[Function]}
    location=""
    params={Object {}}
    path=""
    push={[Function]}
    replace={[Function]}
    url=""
  >
    <Connect(AuthenicatedRoute)
      component={[Function]}
      path="some_path"
      redirectOnLogin="/"
    >
      <AuthenicatedRoute
        component={[Function]}
        dispatch={[Function]}
        path="some_path"
        redirectOnLogin="/"
        session={
          Object {
            "authenticated": true,
          }
        }
      >
        <Route
          component={[Function]}
          path="some_path"
        />
      </AuthenicatedRoute>
    </Connect(AuthenicatedRoute)>
  </MockRouter>
</Provider>
`;

exports[`Component: Authenticated Route renders properly when session not authenicated 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MockRouter
    createHref={[Function]}
    location=""
    params={Object {}}
    path=""
    push={[Function]}
    replace={[Function]}
    url=""
  >
    <Connect(AuthenicatedRoute)
      component={[Function]}
      path="some_path"
      redirectOnLogin="/"
    >
      <AuthenicatedRoute
        component={[Function]}
        dispatch={[Function]}
        path="some_path"
        redirectOnLogin="/"
        session={
          Object {
            "authenticated": false,
          }
        }
      >
        <Redirect
          push={false}
          to={
            Object {
              "pathname": "/login",
              "state": Object {
                "from": "/",
              },
            }
          }
        />
      </AuthenicatedRoute>
    </Connect(AuthenicatedRoute)>
  </MockRouter>
</Provider>
`;
